Separate loader and playground builds

skidding opened the pull request:
Fixes #360

skidding commented:
@reinei you might want to take a look. This should not longer require you to [hack server.js to make it work for your setup](https://github.com/react-cosmos/react-cosmos/issues/297#issuecomment-284510403). Thanks!

jlc467 commented on packages/react-cosmos/src/components/Loader.jsx:
```diff
@@ -127,20 +161,24 @@ class Loader extends React.Component {
         nextProxy={firstProxy.next()}
         component={components[component]}
-        fixture={merge(fixture.unserializable, fixture.serializable)}
+        fixture={merge({}, unserializable, serializable)}
```
👍 Thanks for fixing that.

jlc467 commented:
@skidding Checked out your branch and tried some examples, as well as the cosmos export -- all good! Once this makes it into a beta, I can start testing it everyday.

skidding commented:
Update: for those of us watching this issue, I did not stop working on it. On the contrary, I went two extra levels deep to fix long standing issues and clean up ancient code. Since we're doing major changes, I figured now or never.

That said, here are the main areas I've refacored
- Completely rewrote Loader. There's now Loader and RemoteLoader. The former is used for testing purposes (more on this later) and receives component and fixture via props. The latter is meant to be controlled remotely via window.postMessage (i.e. by Playground)
- Completely rewrote Playground. Changing the Loader-Playground communication required changes to the Playground, which was always a pain because it was a monolithic component written like 2 years ago, tested wierdly using the old test framework. I've taken it into pieces and rewrote it via latest ES6 & React features, split into a number of smaller components and tested nicely using Jest. **Moreover**, I went the extra mile and added fixtures for CP, which means by the end of this we will work on CP inside CP!

In effect:
- react-cosmos-tree will go away
- Entire Mocha/Karma test suite will go away

WIP:
- Fixture editor now yet added to new CP
- react-cosmos package will also disappear, because it contains both Loader & Playground logic. Most likely we'll have react-cosmos-loader and react-cosmos-playground, each with its own entry point. This also paves the way for non-React loaders.

There are even more goodies included, so stay tuned!

skidding commented:
Guys, it's time! ⌛️

![image](https://user-images.githubusercontent.com/250750/27283954-58184062-54ff-11e7-8ba1-c81a39a2d0dc.png)

Looking at the diff is insane, although I encourage you to scan it for key parts (tests, fixtures, yarn.lock files, etc. can be ignored). Best to explore [the branch](https://github.com/react-cosmos/react-cosmos/tree/360-separate-playground) and try it out locally.

**Warning: The fixture editor has not yet been added to the new Component Playground yet**, but I wanted to unlock other issues like #314 by merging this gianormous branch first. There are two things that I will submit (small) separate PRs for after this:

- Adding FixtureEditor back (it won't take long, but I didn't want to rush it either since I want to add some enhancements along the way)
- Updating the Contributing Guide (although this branch already adds to it)

Here's an overview of what this branch includes:
- **Complete separation between Loader & Playground.** Playground bundle is now served statically so we only have one build after all)
  - No more `react-cosmos` package. Packages are now better defined with `react-cosmos-loader` and `react-cosmos-component-playground`.
- **Complete rewrite of Playground. `yarn start:playground`** will fire up some Playground Inception right there! 🎇)
  - Removed `react-component-tree` dependency, and got rid of a lot of old code.
- **Loader rewrite.** Split between RemoteLoader (using window.postMessage to communicate) and Loader (uses props). The latter is useful for unit test setup ([see CP tests](https://github.com/react-cosmos/react-cosmos/blob/827e34c8632a2d59246a0c8e2d98b86b6ee8a557/packages/react-cosmos-component-playground/src/components/DisplayScreen/__tests__/index.jsx#L16-L19)).
- **No more Mocha/Karma tests**. Old test suite was removed along with old code, which makes for a cleaner repo and faster build times.
  - Coverage dropped because of `collectCoverageFrom` Jest option. It probably increased, but we weren't properly calculating it before by ignoring files that weren't touched at all by tests. The new metric is fine since it includes new/experimental code like exporting or `react-cosmos-telescope`
- Bonus: First key bindings to ComponentPlayground: `s` will focus on the fixture search input and `ESC` will clear & blur it. @flaviusone will like this.

There's probably more than what I mentioned for feel free to ask any questions. Looking forward to feedback. Thanks!

flaviusone commented on packages/react-cosmos-component-playground/src/components/ComponentPlayground/index.jsx:
```diff
@@ -0,0 +1,222 @@
+
+    return (
+      <div key="loader" className={styles.loader}>
+        <StarryBg>
```
It would be nice to not render this all the time in the background (for any loaded component). StarryBg is nice but GPU is still working in the background for no reason:
![image](https://user-images.githubusercontent.com/1254386/27409819-1632a720-56ec-11e7-9430-1abd74beb9bb.png)

skidding replied:
Good catch!

skidding replied:
Done in #380

flaviusone commented on packages/react-cosmos-component-playground/src/components/DisplayScreen/__tests__/index.jsx:
```diff
@@ -0,0 +1,24 @@
+
+  beforeAll(() => {
+    wrapper = shallowLoader(
+      <Loader component={DisplayScreen} fixture={hiThereFixture} />
```
I don't understand why did you chose to render `DisplayScreen` through 'react-cosmos-loader' (and maybe also other components in this PR). The `DisplayScreen` is a pretty basic component independent of the loader and this test will fail if someone breaks the loader. Maybe I'm missing something though.

skidding replied:
We're using the Loader as a test util here. And we're importing the compiled `lib`, not the source. `Loader` is similar to `ComponentTree` back in the day. It allows us to reuse fixtures as test setup. This is a simple scenario indeed but I wanted to use it consistently in all Component Playground tests to showcase its use internally ("eating our own dog food"). It becomes more useful when we use something like a Redux fixture to create a complex test setup in one line.

flaviusone commented on packages/react-cosmos-component-playground/src/components/FixtureList/index.jsx:
```diff
@@ -0,0 +1,157 @@
+      // we want to show the correct result.
+      return (
+        match(componentAndFixture, searchText).length !== 0 ||
+        match(fixtureAndComponent, searchText).length !== 0
```
I know this is old code but I vote that we remove matching for `fixtureAndComponent`. 

In time I noticed that this doesn't work that well as it won't prioritize your specific search and the result you want will be further down the list. Also one-way matching (componentAndFixture) is how all major editors work (Sublime, VS Code etc).

I don't know if this change will break a lot of tests but I it's a quick win I think we should do it. WDYT?

skidding replied:
Agreed, I also find it confusing. Also fixed in #380

flaviusone commented on packages/react-cosmos-component-playground/src/components/StarryBg/index.jsx:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+export default StarryBg;
```
This kinda does the same thing as `DisplayScreen` (loads some children) but it doesn't seem to have any tests even if Jest coverage reports 100% (but that's because of the way coverage works). 

Does anything stop us from adding Jest Snapshots?

skidding replied:
Added in #380

skidding commented:
Merged & published under `react-cosmos-webpack@2.0.0-beta.16` :tada: :sparkler:
